src/bin/compiler.ml

```diff
@@ -0,0 +1,116 @@
+open Core_kernel
+open Stanc_mir
+open Stanc_backend
+open Stanc_frontend
+open Stanc_optimization
+
+type 'out verbose = 
+    { ast : Ast.untyped_program option
+    ; typed_ast : Ast.typed_program option 
+    ; mir : Mir.typed_prog option
+    ; optimized_mir : Mir.typed_prog option 
+    ; out : 'out option
+    }
+
+type error = 
+    | Lexing of string * Ast.location
+    | Include of string * Ast.location
+    | Parsing of string * Ast.location_span
+    | SemanticError of string * Ast.location_span
+
+let error_of_parse_error = function 
+    | Errors.Lexing(err,loc) -> Lexing(err,loc)
+    | Include(err,loc) -> Include(err,loc)
+    | Parsing(err,locspan) -> Parsing(err,locspan)
+
+let error_of_semantic_error (err,locspan) = 
+    SemanticError(err,locspan)
+
+
+module type SIG = sig
+    type out
+
+    val ast : string -> (Ast.untyped_program,error) Result.t
+
+    val typed_ast : Ast.untyped_program -> (Ast.typed_program,error) Result.t
+
+    val mir : string -> Ast.typed_program -> Mir.typed_prog
+
+    val compile: string -> (out,error) Result.t
+
+    val compile_verbose: string -> (out verbose,out verbose * error) Result.t
+
+end 
+
+module Make
+    (Backend: Backend.SIG) 
+    (Optimize: Optimization.SIG): SIG 
```

Comment:
Can you lay out the use case for pluggable optimizers like this? From what I can tell we'd best case want to incorporate someone else's optimization but we'd want to choose very specifically where to call it within our optimization pass, no?